Support for sql_exporter with pgMonitor extension (#361)
CrunchyData/pgmonitor: hugo/content/exporter/_index.md

| node_exporter/linux/crunchy-node-exporter-service-rhel.conf | /etc/systemd/system/node_exporter.service.d/crunchy-node-exporter-service-rhel.conf  |
| node_exporter/linux/sysconfig.node_exporter | /etc/sysconfig/node_exporter |

##### sql_exporter

> The sql_exportor takes advantage of the Crunchy pgmonitor-extension (https://github.com/CrunchyData/pgmonitor-extension) to provide a much easier configuration and setup. The extension takes care of creating all the necessary objects inside the database. 

Reviewer: ```suggestion
sql_exporter takes advantage of the Crunchy Data pgmonitor-extension (https://github.com/CrunchyData/pgmonitor-extension) to provide a much easier configuration and setup. The extension takes care of creating all the necessary objects inside the database. 
```

If you need to modify them, see the notes in the files for more details and recommendations:
- {{< shell >}}/etc/systemd/system/node_exporter.service.d/crunchy-node-exporter-service-rhel{{< /shell >}}
- {{< shell >}}/etc/sysconfig/node_exporter{{< /shell >}}
- {{< shell >}}/etc/sysconfig/sql_exporter{{< /shell >}}

#### Database Configuration

##### General Configuration

First, make sure you have installed the PostgreSQL contrib modules.  An example of installing this on a RHEL system would be:

```bash
sudo yum install postgresql##-contrib
```

Where `##` corresponds to your current PostgreSQL version.  

> You will need to modify your {{< shell >}}postgresql.conf{{< /shell >}} configuration file to tell PostgreSQL to load shared libraries.

Reviewer: ```suggestion
You will need to modify your {{< shell >}}postgresql.conf{{< /shell >}} configuration file to tell PostgreSQL to load the following shared libraries:
```

| crunchy_bloat_check_collector.yml              | Collection file with pg_bloat_check queries and metrics |
| crunchy_global_collector.yml                   | Collection file with global level queries and metrics |
| crunchy_per_db_collector.yml                   | Collection file with general per-database level queries and metrics |
| crunchy_pgbouncer_collector_121.yml            | Collection file with pgBouncer queries and metrics for a minimum version of 1.21 |
| crunchy_pg_stat_statements_collector.yml       | Collection file with pg_stat_statements queries and metrics
| crunchy_pg_stat_statements_reset_collector.yml | Collection file with options to allow resetting of pg_stat_statements metrics |

Run the setup_db.sql file on all databases that will be monitored by pgMonitor. At minimum this must be at least the global database so the necessary database objects are created. The `pgmonitor-extension` is expected to be available to be installed in the target database(s) when running this file. Note the setup.sql file is a convenience file and the steps contained within it can be done manually and customized as needed.

> The `sql_exporter.yml.example` file should be copied and renamed to `sql_exporter.yml` since this is what the sysconfig file is expecting to find. This file contains setting for sql_exporter, the list of collection files to use and the configuration for which databases to connect to and which collections to run on each database. Please see the examples inside the file and refer to the upstream project for all of the configuration options available. At the example shows how to run both the global and per-db collections on the default postgres database. It also shows how you can connect to pgBouncer to collect metrics directly from it as well. The collector names that can be used can be found inside the collection files at the top.

Reviewer: ```suggestion
The `sql_exporter.yml.example` file should be copied and renamed to `sql_exporter.yml` since this is what the sysconfig file is expecting to find. This file contains settings for sql_exporter, the list of collection files to use, and the configuration for which databases to connect to and which collections to run on each database. Please see the examples inside the file and refer to the upstream project for all of the configuration options available. The example shows how to run both the global and per-db collections on the default 'postgres' database. It also shows how you can connect to pgBouncer to collect metrics directly from it as well. The collector names that can be used can be found inside the collection files at the top.
```

> Note that your pg_hba.conf will have to be configured to allow the {{< shell >}}ccp_monitoring{{< /shell >}} system user to connect as the {{< shell >}}ccp_monitoring{{< /shell >}} role to any database in the instance. The sql_exporter is set to connect via the local tcp loopback by default. If passwordless login is desired, a .pgpass file can be created for the ccp_monitoring user or the connection configuration can be changed to use a local socket and peer-based authentication can be done instead.

Reviewer: ```suggestion
Note that your pg_hba.conf will have to be configured to allow the {{< shell >}}ccp_monitoring{{< /shell >}} system user to connect as the {{< shell >}}ccp_monitoring{{< /shell >}} role to any database in the instance. sql_exporter is set to connect via the local TCP loopback by default. If passwordless login is desired, a .pgpass file can be created for the ccp_monitoring user or the connection configuration can be changed to use a local socket and peer-based authentication can be done instead.
```

##### PGBouncer

> It is possible for the sql_exporter to connect directly to pgBouncer to collect metrics. Specific settings must be used and the example sql_exporter configuration and relevant collection file(s) have these settings enabled. Please refer to those files.

Reviewer: ```suggestion
It is possible for sql_exporter to connect directly to pgBouncer to collect metrics. Specific settings must be used and the example sql_exporter configuration and relevant collection file(s) have these settings enabled. Please refer to those files.
```

```
jobs:
  - job_name: global_targets
    collectors: [crunchy_global]
    static_configs:
        - targets:
            pg1: 'pg://ccp_monitoring@127.0.0.1:5432/postgres?sslmode=disable'
  - job_name: per_db_targets
    collectors: [crunchy_per_db]
    static_configs:
        - targets:
            postgres: 'pg://ccp_monitoring@127.0.0.1:5432/postgres?sslmode=disable'
            alpha: 'pg://ccp_monitoring@127.0.0.1:5432/alpha?sslmode=disable'
            beta: 'pg://ccp_monitoring@127.0.0.1:5432/beta?sslmode=disable'
            charlie: 'pg://ccp_monitoring@127.0.0.1:5432/charlie?sslmode=disable'
```

## Metrics Collected {#metrics-collected}

The metrics collected by our exporters are outlined below. 

### PostgreSQL {#postgresql}

> PostgreSQL metrics are collected by the [sql_exporter](https://github.com/burningalchemist/sql_exporter). pgMonitor uses custom queries for its PG metrics. 

Reviewer: ```suggestion
PostgreSQL metrics are collected by [sql_exporter](https://github.com/burningalchemist/sql_exporter). pgMonitor uses custom queries for its PG metrics. 
```

#### pg_stat_statements Metrics

> Collecting all per-query metrics into Prometheus could greatly increase storage requirements and heavily impact performance without sufficient resources. Therefore the metrics below give simplified numeric metrics on overall statistics and Top N queries. N is set as the LIMIT value in the `crunchy_pg_stat_statements_collector.yml` collections file. If you would like to adjust this number, it is recommended to make a copy of this collection file and use that modified file in your sql_exporter collector file config instead.

Reviewer: ```suggestion
Collecting all per-query metrics into Prometheus could greatly increase storage requirements and heavily impact performance. Therefore, the metrics below give simplified numeric metrics on overall statistics and Top N queries. N is set as the LIMIT value in the `crunchy_pg_stat_statements_collector.yml` collections file. If you would like to adjust this number, it is recommended to make a copy of this collection file and use that modified file in your sql_exporter collector file config instead.
```

> . Note that the statistics for individual queries can only be reset on PG12+. Prior to that, pg_stat_statements must have all statistics reset to redo the top N queries.

Reviewer: ```suggestion
Note that the statistics for individual queries can only be reset on PG12+. Prior to that, pg_stat_statements must have all statistics reset to redo the top N queries.
```

### System {#system}

\*NIX Operating System metrics (Linux, BSD, etc) are collected using the [node_exporter](https://github.com/prometheus/node_exporter) provided by the Prometheus team. pgMonitor only collects the default metrics provided by node_exporter, but many additional metrics are available if needed.

## Legacy postgres_exporter Setup {#postgres-exporter}

> If you had been using pgMonitor prior to version 5.0.0, postgres_exporter was the method used to collect PostgreSQL metrics. This exporter can still be used with 5.0.0, but there are some additional steps required. It is HIGHLY recommended to switch to using sql_exporter as soon as possible. Custom query support will be dropped from postgres_exporter and no new features of pgMonitor are being developed around it.

Reviewer: You mention that custom query support is being dropped, but you don't explain that that would completely break pgM since it relies on it. So I'd suggest either removing that comment in the multiple places you have it, or expand it to explain why it matters. (If it were me, I'd just drop it everywhere)

Author: I updated this to explain why

|       Column       |       Description                                                |
|--------------------|------------------------------------------------------------------|
| view_schema     | Schema containing the materialized view |
| view_name       | Name of the materialized view |
| concurrent_refresh | Boolean that sets whether this materialized view can be refreshed concurrently (requires a unique index) |
| run_interval       | How often this materialized view should have its data refreshed. Must be a value compatible with the PG interval type   |
| last_run           | Timestamp of the last time this view was refreshed |
| active             | Boolean that sets whether this view should be refreshed when the procedure is called |
| scope              | Whether the data contained in the view is per-database or instance-wide. Currently unused |

You are also free to use this materialized view system for your own custom metrics as well. Simply make a materialized view, add its name to the configuration table and ensure the user running the refresh has permissions to do so for your view(s).

##### PGBouncer

> In order to monitor pgbouncer with postgres_exporter, the pgbouncer_fdw maintained by CrunchyData is required. Please see its repository for full installation instructions. A package for this is available for Crunchy customers.

Reviewer: proper case pgbouncer in this line and line 593?

also shouldn't it be 'Crunchy Data customers' ?

Author: fixed

https://github.com/CrunchyData/pgbouncer_fdw

Once that is working, you should be able to add the {{< shell >}}queries_pgbouncer.yml{{< /shell >}} file to the {{< yaml >}}QUERY_FILE_LIST{{< /shell >}} for the exporter that is monitoring the database where the FDW was installed.

> #### Monitoring multiple databases and/or running multiple postgres exporters (RHEL / CentOS)

Reviewer: still want to list CentOS in this release since it dies soon? or remove it in a later release?

Author: I'll just remove it for now

Certain metrics are not cluster-wide, so multiple exporters must be run to avoid duplication when monitoring multiple databases in a single PostgreSQL instance. To collect these per-database metrics, an additional exporter service is required and pgMonitor provides this using the following query file: ({{< shell >}}queries_per_db.yml{{< /shell >}}). In Prometheus, you can then define the global and per-db exporter targets for a single job. This will place all the metrics that are collected for a single database instance together.

{{< note >}}The "setup.sql" file does not need to be run on these additional databases if using the queries that pgMonitor comes with.{{< /note >}}

> pgMonitor provides and recommends an example sysconfig file for this per-db exporter: {{< shell >}}sysconfig.postgres_exporter_pg##_per_db{{< /shell >}}. If you'd like to create additional exporter services for different query files, just copy the existing ones and modify the relevant lines, mainly the port, database name, and query file. The below example shows connecting to 3 databases in the same instance to collect their per-db metrics: `postgres`, `mydb1`, and `mydb2`.

Reviewer: ```suggestion
pgMonitor provides and recommends an example sysconfig file for this per-db exporter: {{< shell >}}sysconfig.postgres_exporter_pg##_per_db{{< /shell >}}. If you'd like to create additional exporter services for different query files, just copy the existing ones and modify the relevant lines, mainly the port, database name, and query file. The below example shows connecting to three databases in the same instance to collect their per-db metrics: `postgres`, `mydb1`, and `mydb2`.
```
